mt_gateway_test.go

```diff
@@ -0,0 +1,369 @@
+	require.NoError(t, err)
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		Run(ctx)
+	}()
+
+	serviceHealthEndpoint := fmt.Sprintf("http://localhost:%d/health", httpPort)
+	t.Log("serviceHealthEndpoint", serviceHealthEndpoint)
+	health.WaitUntilReady(ctx, t,
+		serviceHealthEndpoint,
+		time.Minute,
+		250*time.Millisecond,
+		"serviceHealthEndpoint",
+	)
+
+	select {
+	case ack := <-etcdContainer.Client.Watch(ctx, "test-ack/1"):
+		v, err := unmarshalWorkspaceAckValue(t, &ack)
+		require.NoError(t, err)
+		require.Equal(t, "RELOADED", v.Status)
+		require.Equal(t, "", v.Error)
+	case <-time.After(20 * time.Second):
+		t.Fatal("Timeout waiting for test-ack/1")
+	}
+
+	require.Empty(t, webhook.Requests(), "webhook should have no requests before sending the events")
+	sendEventsToGateway(t, httpPort, writeKey)
+
+	var (
+		eventPayload string
+		message      map[string]interface{}
+	)
+	require.Eventually(t, func() bool {
+		return postgresContainer.DB.QueryRowContext(ctx,
+			"SELECT event_payload FROM gw_jobs_1 WHERE workspace_id = $1", workspaceID,
+		).Scan(&eventPayload) == nil
+	}, time.Minute, 50*time.Millisecond)
+	require.NoError(t, json.Unmarshal([]byte(eventPayload), &message))
+
+	batch, ok := message["batch"].([]interface{})
+	require.True(t, ok)
+	require.Len(t, batch, 1)
+	require.Equal(t, message["writeKey"], writeKey)
+	for _, msg := range batch {
+		m, ok := msg.(map[string]interface{})
+		require.True(t, ok)
+		require.Equal(t, "anonymousId_1", m["anonymousId"])
+		require.Equal(t, "identified_user_id", m["userId"])
+		require.Equal(t, "identify", m["type"])
+		require.Equal(t, "1", m["eventOrderNo"])
+		require.Equal(t, "messageId_1", m["messageId"])
+	}
+
+	// Only the Gateway is running, so we don't expect any destinations to be hit.
+	require.EqualValues(t, 0, webhook.RequestsCount(), "webhook should have no requests because there is no processor")
+
+	serverModeReqKey := getETCDServerModeReqKey(t, serverInstanceID)
+	t.Logf("Server mode ETCD key: %s", serverModeReqKey)
+
+	{ // Trigger degraded mode, the Gateway should still work
+		_, err = etcdContainer.Client.Put(ctx, serverModeReqKey, `{"mode":"DEGRADED","ack_key":"test-ack/2"}`)
+		require.NoError(t, err)
+		select {
+		case ack := <-etcdContainer.Client.Watch(ctx, "test-ack/", clientv3.WithPrefix()):
+			t.Logf("ACK: %+v", ack)
+		case <-time.After(20 * time.Second):
+			t.Fatal("Timeout waiting for server-mode test-ack")
+		}
+
+		sendEventsToGateway(t, httpPort, writeKey)
+		require.Eventually(t, func() bool {
+			var count int
+			err := postgresContainer.DB.QueryRowContext(ctx,
+				"SELECT COUNT(*) FROM gw_jobs_1 WHERE workspace_id = $1", workspaceID,
+			).Scan(&count)
+			if err != nil {
+				return false
+			}
+			return count == 2
+		}, time.Minute, 50*time.Millisecond)
+		require.NoError(t, json.Unmarshal([]byte(eventPayload), &message))
+	}
+
+	{ // Checking that Workspace Changes errors are handled
```

Comment:
What shall happen if you send an event for a workspace that is not part of the node's workspaceIDs? Shall the gateway accept it?

Reply:
The gateway is quite dumb in this regard and I think it's fine for now. The one that knows about this is the proxy if I'm not mistaken where we keep a `writeKey` to workspace Map in the form of a Lua table. If there is no entry in the map then we reject it at the proxy level.

This is mostly what happens with IAM services as well. Requests coming from the outside can hit only "exposed" services (e.g. IAM service) that authenticate the request and let it through to the internal services only if it's valid. Then the internal services can avoid having the validation/authentication logic copied/duplicated in each µsvc because they can be hit from internal calls only so they can trust the requests.

I think the only case when the GW can be hit from outside is for self-hosted (open-source) deployments (cc @cisse21).

Reply:
The actual question was mostly referring to the scenario of a request with a valid workspaceID (writeKey), but not belonging to the node's list of workspaceIDs.
I understand that since the node keeps in memory the configuration from all workspaces, it will accept the request and store the event in the `gw` table, even though this is an event for a workspace not assigned to it.

Reply:
I see what you mean and although useful to help detect misbehaviour it still feels a bit like throw away code at this time, given that we haven't taken a decision on whether a node should fetch the configuration for all WS or not. I'd like to get back to this once I know more about that aspect if you don't mind.

Reply:
One could argue that if `rudder-server` doesn't respect the list of workspaceIDs allotted to it, there is no actual reason for keeping this list in `rudder-server`

Reply:
I'm setting up a call to decide what we want RS to do.

Reply:
We decided to simplify and have it fetch all workspaces configuration till the protocol is at least a bit more mature.